core/src/main/java/com/scalar/db/api/ScanAll.java

```diff
@@ -0,0 +1,101 @@
+package com.scalar.db.api;
+
+import com.google.common.base.MoreObjects;
+import com.scalar.db.io.Key;
+import java.util.Collection;
+import java.util.Objects;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A command to retrieve all the entries of the database. The scan range of clustering key and
+ * {@link Ordering} cannot be specified for this command. The number of {@link Result} can be
+ * limited.
+ */
+@NotThreadSafe
+public class ScanAll extends Scan {
+
+  private static final Key DUMMY_PARTITION_KEY = Key.of();
+
+  public ScanAll() {
+    super(DUMMY_PARTITION_KEY);
+  }
+
+  public ScanAll(ScanAll scanAll) {
+    super(scanAll);
+  }
+
+  @Override
+  public ScanAll withStart(Key clusteringKey) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public ScanAll withStart(Key clusteringKey, boolean inclusive) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Scan withEnd(Key clusteringKey) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Scan withEnd(Key clusteringKey, boolean inclusive) {
+    throw new UnsupportedOperationException();
+  }
```

Comment:
If I understood correctly, you would like to implement these methods for ScanAll later on. 
Though intuitively, as a user, If I want only to retrieve a subset of all the database entries, wouldn't I want to filter on the entire primary key and not only on the clustering key only?

Reply:
In this `ScanAll` operation, we won't support a clustering key range scan. Instead, we can add other methods filtering on arbitrary columns in the future. At this point, we support only specifying projections and limit.

Reply:
All right. 
I wouldn't say I like that we have to expose the non-implemented `withEnd` and `withStart` methods to the end-user, but if it's a necessary evil to reduce the development cost, then I understand.